packages/toolkit/src/query/core/buildMiddleware/batchActions.ts

```diff
@@ -14,39 +16,75 @@ const queueMicrotaskShim =
+  const subscriptionsPrefix = `${api.reducerPath}/subscriptions`
+
+  let previousSubscriptions: SubscriptionState =
+    null as unknown as SubscriptionState
+
   let dispatchQueued = false
 
-  return (action, mwApi) => {
-    if (queryThunk.rejected.match(action)) {
-      const { condition, arg } = action.meta
-
-      if (condition && arg.subscribe) {
-        // request was aborted due to condition (another query already running)
-        // _Don't_ dispatch right away - queue it for a debounced grouped dispatch
-        abortedQueryActionsQueue.push(action)
-
-        if (!dispatchQueued) {
-          queueMicrotaskShim(() => {
-            mwApi.dispatch(
-              api.internalActions.subscriptionRequestsRejected(
-                abortedQueryActionsQueue
-              )
-            )
-            abortedQueryActionsQueue = []
-            dispatchQueued = false
-          })
-          dispatchQueued = true
-        }
-        // _Don't_ let the action reach the reducers now!
-        return false
+  return (action, mwApi, internalState) => {
+    if (!previousSubscriptions) {
+      // Initialize it the first time this handler runs
+      previousSubscriptions = JSON.parse(
+        JSON.stringify(internalState.currentSubscriptions)
+      )
+    }
+
+    // Intercept requests by hooks to see if they're subscribed
+    // Necessary because we delay updating store state to the end of the tick
+    if (api.internalActions.internal_probeSubscription.match(action)) {
+      const { queryCacheKey, requestId } = action.payload
+      const hasSubscription =
+        !!internalState.currentSubscriptions[queryCacheKey]?.[requestId]
+      return [false, hasSubscription]
+    }
+
+    // Update subscription data based on this action
+    const didMutate = actuallyMutateSubscriptions(
+      internalState.currentSubscriptions,
+      action
+    )
+
+    if (didMutate) {
+      if (!dispatchQueued) {
+        queueMicrotaskShim(() => {
+          // Deep clone the current subscription data
+          const newSubscriptions: SubscriptionState = JSON.parse(
+            JSON.stringify(internalState.currentSubscriptions)
+          )
+          // Figure out a smaller diff between original and current
+          const [, patches] = produceWithPatches(
+            previousSubscriptions,
+            () => newSubscriptions
+          )
+
+          // Sync the store state for visibility
+          mwApi.next(api.internalActions.subscriptionsUpdated(patches))
+          // Save the cloned state for later reference
+          previousSubscriptions = newSubscriptions
+          dispatchQueued = false
+        })
+        dispatchQueued = true
       }
+
+      const isSubscriptionSliceAction =
+        !!action.type?.startsWith(subscriptionsPrefix)
+      const isAdditionalSubscriptionAction =
+        queryThunk.rejected.match(action) &&
+        action.meta.condition &&
+        !!action.meta.arg.subscribe
+
+      const actionShouldContinue =
```

Comment:
I still debate the "breakingness" of it. Sure, those actions were publicly visible, but I still see that as an implementation detail, not something we expect users to rely on.  (Which of course means that someone somewhere probably _is_ relying on them...)

My fear is that as soon as we start letting all those actions through we lose any perf improvement we've gotten out of this effort, because we're right back to running reducers and notifying subscribers each time.

Reply:
If the actions don't actually change any state we won't notify subscribers though, right?

Reply:
Subscribers are _always_ notified, after _every_ action.

They normally do have an early bailout, like `if (newState !== prevState) return`, but that's still a bunch of unnecessary calls.  Similar for the reducers - they may no-op out, but they still had to get called.

Reply:
Ah, I somehow had convinced myself that there was a bailout before calling the subscribers (since the subscribers have no access to the action, calling them without a state change seemed unreasonable). In that case, yeah, that's a point for skipping it. I'd still prefer keeping them until 2.0, but I can also live with them going away :)